src/levels.rs

```diff
@@ -1,5 +1,1053 @@
+            }
+        }
+
+        if opts.in_memory {
+            return Ok(core);
+        }
+
+        // TODO: Revert to manifest.
+
+        let manifest_data = manifest.manifest_cloned();
+
+        let mut max_file_id = 0;
+        let mut tables: Vec<Vec<Table>> = Vec::new();
+        tables.resize(opts.max_levels, vec![]);
+
+        // TODO: Parallelly open tables.
+        for (id, table_manifest) in manifest_data.tables {
+            if id > max_file_id {
+                max_file_id = id;
+            }
+
+            let table_opts = build_table_options(&opts);
+            // TODO: Set compression, data_key, cache.
+
+            let filename = crate::table::new_filename(id, &opts.dir);
+            let table = Table::open(&filename, table_opts)?;
+            // TODO: Allow checksum mismatch tables.
+
+            tables[table_manifest.level as usize].push(table);
+        }
+
+        core.next_file_id.store(max_file_id + 1, Ordering::SeqCst);
+
+        for (i, tables) in tables.into_iter().enumerate() {
+            core.levels[i].write().unwrap().init_tables(tables);
+        }
+
+        // TODO: Validate core.
+
+        util::sync_dir(&opts.dir)?;
+
+        Ok(core)
+    }
+
+    /// Calculates the [`Targets`] for levels in the LSM tree.
+    fn level_targets(&self) -> Targets {
+        let adjust = |size| {
+            if size < self.opts.base_level_size {
+                self.opts.base_level_size
+            } else {
+                size
+            }
+        };
+
+        let mut targets = Targets {
+            base_level: 0,
+            target_size: vec![0; self.levels.len()],
+            file_size: vec![0; self.levels.len()],
+        };
+
+        // db_size is the size of the last level.
+        let mut db_size = self.last_level().read().unwrap().total_size;
+
+        for i in (1..self.levels.len()).rev() {
+            let ltarget = adjust(db_size);
+            targets.target_size[i] = ltarget;
+            if targets.base_level == 0 && ltarget <= self.opts.base_level_size {
+                targets.base_level = i;
+            }
+            db_size /= self.opts.level_size_multiplier as u64;
+        }
+
+        let mut tsz = self.opts.base_level_size;
+
+        // Use mem_table_size for Level 0. Because at Level 0, we stop
+        // compactions based on the number of tables, not the size of
+        // the level. So, having a 1:1 size ratio between memtable size
+        // and the size of L0 files is better than churning out 32 files
+        // per memtable (assuming 64MB mem_table_size and 2MB base_table_size).
+        for i in 0..self.levels.len() {
+            if i == 0 {
+                targets.file_size[i] = self.opts.mem_table_size;
+            } else if i <= targets.base_level {
+                targets.file_size[i] = tsz;
+            } else {
+                tsz *= self.opts.table_size_multiplier as u64;
```

Comment:
Why not multiple with level_size_multiplier here?

Reply:
Because here we calculate `file_size`, which indicates the size of SSTable, rather than the size of each level.